---
title: "Preprocessing of Notch and Cph RNAi samples"
author: "Nick Hirschm√ºller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F)
set.seed(123)
```


```{r load libraries}
library(here)
library(DropletUtils)
library(tidyverse)
library(biomaRt)
library(scuttle)
library(scales)
library(patchwork)

# load custom functions
source(here("plot_theme.R"))
```


## Introduction
Starting with the raw, unfiltered data that we get from running CellRanger, we will identify empty droplets and remove cells that do not pass our quality control.


### Load Data
```{r}
# Cell ranger outputs are separated by sample ID
sample_ids <- c("TX49_rep1", "TX49_rep2") # this is our internal ID
file_names <- c("DS-00248", "DS-00249")


# Loop through the directories and load data
raw_sce <- lapply(setNames(seq_along(file_names), sample_ids), function(index) {
    sample <- sample_ids[index]
    file <- file_names[index]

    # read in the 10X as a SingleCellExperiment
    tmp <- read10xCounts(
        samples = here("raw_data", "NotchCphRNAi", file, "raw_feature_bc_matrix"),
        sample.names = sample
    )

    # we can remove some unnecessary columns to save memory
    rowData(tmp) <- rowData(tmp)[, "Symbol"]
    colnames(rowData(tmp)) <- "Symbol"
    return(tmp)
})
```

### Add gene annotations
```{r}
# load in some additional gene information which was fetched using biomart.
# ensembl <- useMart("ensembl")
# ensembl <- useDataset("dmelanogaster_gene_ensembl",mart=ensembl)
#
# attributes_of_interest <- c("flybase_gene_id","description",
#              "chromosome_name","start_position","end_position")
#
# gene_mdata <- getBM(attributes=attributes_of_interest,
#       filters = 'flybase_gene_id',
#       values = all_genes %>% unlist() %>% unique(),
#       mart = ensembl)
#
# gene_mdata %>% data.table::fwrite(., file=here("raw_data", "gene_annotation.tsv"),sep="\t", row.names = F)
gene_mdata <- data.table::fread(here("raw_data", "gene_annotation.tsv"))

# add the additional data to the rowData
raw_sce <- lapply(setNames(sample_ids, sample_ids), function(sample) {
    tmp <- raw_sce[[sample]]

    rowData(tmp) <- rowData(tmp) %>%
        as.data.frame() %>%
        rownames_to_column("flybase_gene_id") %>%
        left_join(., gene_mdata) %>%
        mutate_all(~ replace(., is.na(.), "")) %>% # replace NA with empty value
        column_to_rownames("flybase_gene_id")

    return(tmp)
})

# check out which genes are mitochondrial (marker for dying cells)
mitochondrial_genes <- lapply(setNames(sample_ids, sample_ids), function(sample) {
    rowData(raw_sce[[sample]])$chromosome_name == "mitochondrion_genome"
})
```

### Identify empty droplets
We use the knee plot for that.

Look at the knee plot and determine a suitable cutoff
```{r}
options(warn = 1)
raw_sce <- lapply(setNames(sample_ids, sample_ids), function(sample) {
    tmp <- raw_sce[[sample]]
    tmp <- addPerCellQC(tmp, subsets = list(Mito = mitochondrial_genes[[sample]]))
    tmp <- addPerFeatureQC(tmp)
    return(tmp)
})

knee_df <- lapply(setNames(sample_ids, sample_ids), function(sample) {
    tmp <- raw_sce[[sample]]

    knee_df <- barcodeRanks(tmp) %>%
        data.frame() %>%
        arrange(rank)
})


## Cell ranger uses 500 as the cutoff.

# for tx47
knee_plot_tx47 <- ggplot(knee_df$TX49_rep1, aes(x = rank, y = total)) +
    geom_line() +
    geom_hline(yintercept = 750, linetype = "dotted", col = "red") +
    annotate("text", x = 30, y = 750, label = "Inflection Point", vjust = -0.5, size = 3.3) +
    scale_x_log10(
        breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))
    ) +
    scale_y_log10(
        breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))
    ) +
    annotation_logticks() +
    theme_Publication() +
    ggtitle("TX49_rep1")

# for tx49
knee_plot_tx49 <- ggplot(knee_df$TX49_rep2, aes(x = rank, y = total)) +
    geom_line() +
    geom_hline(yintercept = 750, linetype = "dotted", col = "red") +
    annotate("text", x = 30, y = 750, label = "Inflection Point", vjust = -0.5, size = 3.3) +
    scale_x_log10(
        breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))
    ) +
    scale_y_log10(
        breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))
    ) +
    annotation_logticks() +
    theme_Publication() +
    ggtitle("TX49_rep2")

knee_plot <- knee_plot_tx47 + knee_plot_tx49
knee_plot

# filter based on our determined cutoffs.
raw_sce <- lapply(setNames(sample_ids, sample_ids), function(sample) {
    tmp <- raw_sce[[sample]]
    if (sample == "TX49_rep1") {
        mask <- tmp$sum > 750
    } else {
        mask <- tmp$sum > 750
    }
    tmp <- tmp[, mask]
    return(tmp)
})
```


### Remove low quality cells.
Low quality cells were defined as follows:
1) high content of mitochondrial reads 
2) few detected genes (<250)
3) very high number of UMIs (top 1 percentile)


```{r}
# look at the distribution of mitochondrial reads
mitochondrial_reads_density_plot <- lapply(setNames(sample_ids, sample_ids), function(x) {
    tmp <- raw_sce[[x]]
    ggplot(colData(tmp) %>% data.frame(), aes(x = subsets_Mito_percent)) +
        geom_density() +
        theme_Publication() +
        geom_vline(xintercept = 10)
})


mitochondrial_reads_density_plot$TX49_rep1 +
    ggtitle("TX49_rep1")

mitochondrial_reads_density_plot$TX49_rep2 +
    ggtitle("TX49_rep2")


# use a hard cutoff of 10%
mitochondrial_filter_mask <- list(
    "TX49_rep1" = raw_sce$TX49_rep1$subsets_Mito_percent > 10,
    "TX49_rep2" = raw_sce$TX49_rep2$subsets_Mito_percent > 10
)

gene_filter_mask <- lapply(setNames(sample_ids, sample_ids), function(x) {
    tmp <- raw_sce[[x]]
    tmp$detected < 250
})

umi_filter_mask <- lapply(setNames(sample_ids, sample_ids), function(x) {
    tmp <- raw_sce[[x]]
    tmp$total > quantile(tmp$total, 0.99)
})

# combine all masks
discard <- lapply(setNames(sample_ids, sample_ids), function(x) {
    to_discard <- mitochondrial_filter_mask[[x]] | gene_filter_mask[[x]] | umi_filter_mask[[x]]
})

# create a dataframe for an overview why cells got kicked out
reason <- lapply(sample_ids, function(x) {
    data.frame(
        sample = x,
        gene_filter = sum(gene_filter_mask[[x]]),
        umi_filter = sum(umi_filter_mask[[x]]),
        mito_filter = sum(mitochondrial_filter_mask[[x]]),
        all = sum(discard[[x]])
    )
}) %>% bind_rows()

# add a column to the sce to  indicate which cells we will discard
raw_sce <- lapply(setNames(sample_ids, sample_ids), function(x) {
    tmp <- raw_sce[[x]]
    tmp$discard <- discard[[x]]
    return(tmp)
})

# perform the filtering
sce_qc <- lapply(setNames(sample_ids, sample_ids), function(x) {
    tmp <- raw_sce[[x]]
    tmp <- tmp[, !colData(tmp)$discard]
    return(tmp)
})

# save the filtered object
saveRDS(sce_qc, here("output", "NotchCphRNAi_qced.rds"))
```


```{r}
sessionInfo()
```







