---
title: "Figure out which genes are under the influence of Cph"
author: "Nick Hirschm√ºller"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F)
set.seed(123)
```-


```{r}
library(Seurat)
library(tidyverse)
library(patchwork)
library(here)
library(data.table)

source(here("plot_theme.R"))
source(here("helper_functions.R"))
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, warning = F, message = F)
set.seed(123)
```


### Introduction
We will use the information from the single cell sequencing and also from the NanoDam experiment.
We want to identify genes that are upregulated in NotchRNAi and those whose expression goes back to WT level when Cph is also knocked down.
I.e. Identify genes upregulated in NotchRNAi vs ctrl and also in NotchRNAi vs NotchRNAi+CphRNAi.
Then check which of these genes are targets of Cph in the nano dam data.

```{r}
seurat <- readRDS(here("output", "Ctrl_Notch_NotchCphRNAi_integrated_scent.rds"))
progenitors <- seurat[, seurat$high_res_annotation %in% c("EEP", "ISC", "EB")]
DefaultAssay(progenitors) <- "RNA"
# load the MAST differential gene expression results
MAST_notch_ctrl <- fread(here("output", "differential_expression", "MAST", "NotchRNAi", "Progenitor", "MAST_diff_expr_NotchRNAi_Progenitor_coarse.tsv"), data.table = F)

MAST_notch_cphnotch <- fread(here("output", "differential_expression", "MAST", "NotchRNAi_vs_NotchCphRNAi", "Progenitor", "MAST_diff_expr_NotchRNAi_vs_NotchCphRNAi_Progenitor_coarse.tsv"), data.table = F)

# We will substitute the logFC calculated by MAST by a simple comparison of the means (as does Seurat) (https://github.com/satijalab/seurat/blob/b56d194939379460db23380426d3896b54d91ab6/R/differential_expression.R#L553)

# We will actually use seurat for that
Idents(progenitors) <- progenitors$perturbation
seurat_res <- FindMarkers(
    progenitors,
    ident.1 = "NotchRNAi",
    ident.2 = "ctrl",
    features = c(MAST_notch_ctrl$gene),
    logfc.threshold = 0,
    min.pct = 0
) %>% rownames_to_column("gene")
MAST_notch_ctrl <- MAST_notch_ctrl %>%
    left_join(., seurat_res %>% select(gene, avg_log2FC_seurat = avg_log2FC))


seurat_res <- FindMarkers(
    progenitors,
    ident.1 = "NotchRNAi",
    ident.2 = "NotchCphRNAi",
    features = c(MAST_notch_cphnotch$gene),
    logfc.threshold = 0,
    min.pct = 0
) %>% rownames_to_column("gene")
MAST_notch_cphnotch <- MAST_notch_cphnotch %>%
    left_join(., seurat_res %>% select(gene, avg_log2FC_seurat = avg_log2FC))
```


```{r}
notch_ctrl_upregulated <- MAST_notch_ctrl %>%
    filter(avg_log2FC_seurat > 0.25) %>%
    filter(p_adj < 0.05) %>%
    pull(gene)

notch_cphnotch_upregulated <- MAST_notch_cphnotch %>%
    filter(avg_log2FC_seurat > 0.25) %>%
    filter(p_adj < 0.05) %>%
    pull(gene)

# get the overlap
overlap <- intersect(notch_ctrl_upregulated, notch_cphnotch_upregulated)

# load nanodam data
nanodam <- fread(here("raw_data", "diffbind_gene_peak_summary_ranked_exon_promoter.csv"), data.table = F)

nanodam_targets <- nanodam %>%
    filter(FDR < 0.05) %>%
    pull(gene_symbol)

overlap_diff_expr_nanodam <- intersect(overlap, nanodam_targets)
```


# plot these genes
```{r}
seurat <- readRDS(here("output", "Ctrl_Notch_NotchCphRNAi_integrated_scent.rds"))
seurat$tmp_celltype <- "Progenitor"
seurat$perturbation <- factor(seurat$perturbation, levels = c("ctrl", "NotchRNAi", "NotchCphRNAi"))
seurat$dummy <- case_when(
    seurat$perturbation == "ctrl" ~ "A",
    seurat$perturbation == "NotchRNAi" ~ "B",
    seurat$perturbation == "NotchCphRNAi" ~ "C",
)
progenitors <- seurat[, seurat$high_res_annotation %in% c("EEP", "ISC", "EB")]
DefaultAssay(progenitors) <- "RNA"


library(ggtext)
jitter_plot_fun <- function(srt, gene, population) {
    mdata <- srt@meta.data[, c("tmp_celltype", "perturbation", "dummy")]
    exp_data <- FetchData(srt, gene)
    colnames(exp_data) <- gene
    stopifnot(all(rownames(mdata) == rownames(exp_data)))

    plot_data <- cbind(mdata, exp_data)
    p <- ggplot(plot_data, aes(x = tmp_celltype, y = !!sym(gene))) +
        geom_point(mapping = aes(color = perturbation, fill = perturbation), size = 0.35, stroke = 0.2, shape = 21, position = position_jitterdodge(dodge.width = 0.9)) +
        geom_violin(
            mapping = aes(color = dummy),
            position = position_dodge(width = 0.9),
            fill = NA,
            adjust = 1, trim = TRUE,
            scale = "width", show.legend = F,
            width = 0.72 # adjust width so the violine does not stick out too much
        ) +
        theme_Publication_side_legend() %+%
        theme(
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            legend.text = element_markdown(size = 15),
            legend.title = element_text(size = 16)
        ) +
        ylab("Expression\n(logcounts)") +
        guides(
            color = guide_legend(
                title = "Condition",
                override.aes = list(size = 2, alpha = 1, shape = NULL),
                hjust = 0
            ),
            fill = "none"
        ) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.08))) +
        stat_summary(fun = "mean", geom = "point", aes(group = perturbation), color = "black", size = 3, shape = 8, stroke = 0.95, position = position_dodge(width = 0.9)) +
        scale_color_manual(
            values = c("ctrl" = color_mapping[1], "NotchRNAi" = color_mapping[4], NotchCphRNAi = "#00ba38", A = "black", B = "black", C = "black"),
            labels = c(
                ctrl = "Control",
                NotchRNAi = "*Notch*<sup><i>RNAi</i></sup>",
                NotchCphRNAi = "*Cph*<sup><i>RNAi</i></sup>+*Notch*<sup><i>RNAi</i></sup>"
            )
        ) +
        scale_fill_manual(values = alpha(c(color_mapping[1], color_mapping[4], "#00ba38"), 0.4)) +
        ggtitle(paste(population, gene)) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.15)))


    return(p)
}

significance_marker <- function(input) {
    if (input < 0.001) {
        return("***")
    } else if (input < 0.01) {
        return("**")
    } else if (input < 0.05) {
        return("*")
    } else {
        return("")
    }
}


for (g in overlap_diff_expr_nanodam) {
    print(g)
    population <- "progenitors"

    # Create an annotation dataframe
    max_ <- FetchData(progenitors, g)[, 1] %>% max()

    sig_notch_ctrl <- MAST_notch_ctrl %>%
        filter(gene == g) %>%
        pull(p_adj) %>%
        significance_marker(.)

    sig_notch_cphnotch <- MAST_notch_cphnotch %>%
        filter(gene == g) %>%
        pull(p_adj) %>%
        significance_marker(.)

    print(sig_notch_cphnotch)
    print(sig_notch_ctrl)
    annotation_df <- data.frame(
        celltype = c("a", "b"),
        y_position_ = c(max_ + 0.4, max_ + 0.8), # Adjust as needed
        xmin_ = c(0.7, 1),
        xmax_ = c(1, 1.3),
        label = c(sig_notch_ctrl, sig_notch_cphnotch)
    )

    p <- jitter_plot_fun(progenitors, g, population) +
        ggsignif::geom_signif(
            data = annotation_df,
            aes(xmin = xmin_, xmax = xmax_, annotations = label, y_position = y_position_, group = celltype),
            textsize = 6.5, manual = TRUE, vjust = 0,
            inherit.aes = F
        )

    ggsave(
        plot = p,
        filename = here("plots", "genes_under_cph_control", str_interp("${g}_jitter_plot_cph_pos_regulation.pdf")),
        width = 3.5,
        height = 1.8,
        scale = 2
    )
}
```



# Now do the same but in reverse
I.e. find genes that are negatively regulated by Cph

```{r}
notch_ctrl_downregulated <- MAST_notch_ctrl %>%
    filter(avg_log2FC_seurat < -0.25) %>%
    filter(p_adj < 0.05) %>%
    pull(gene)

notch_cphnotch_downregulated <- MAST_notch_cphnotch %>%
    filter(avg_log2FC_seurat < -0.25) %>%
    filter(p_adj < 0.05) %>%
    pull(gene)

# get the overlap
overlap <- intersect(notch_ctrl_downregulated, notch_cphnotch_downregulated)

# load nanodam data
nanodam <- fread(here("raw_data", "diffbind_gene_peak_summary_ranked_exon_promoter.csv"), data.table = F)

nanodam_targets <- nanodam %>%
    filter(FDR < 0.05) %>%
    pull(gene_symbol)

overlap_diff_expr_nanodam <- intersect(overlap, nanodam_targets)
```


# plot these genes
```{r}
seurat <- readRDS(here("output", "Ctrl_Notch_NotchCphRNAi_integrated_scent.rds"))
seurat$tmp_celltype <- "Progenitor"
seurat$perturbation <- factor(seurat$perturbation, levels = c("ctrl", "NotchRNAi", "NotchCphRNAi"))
seurat$dummy <- case_when(
    seurat$perturbation == "ctrl" ~ "A",
    seurat$perturbation == "NotchRNAi" ~ "B",
    seurat$perturbation == "NotchCphRNAi" ~ "C",
)
progenitors <- seurat[, seurat$high_res_annotation %in% c("EEP", "ISC", "EB")]
DefaultAssay(progenitors) <- "RNA"


library(ggtext)
jitter_plot_fun <- function(srt, gene, population) {
    mdata <- srt@meta.data[, c("tmp_celltype", "perturbation", "dummy")]
    exp_data <- FetchData(srt, gene)
    colnames(exp_data) <- gene
    stopifnot(all(rownames(mdata) == rownames(exp_data)))

    plot_data <- cbind(mdata, exp_data)
    p <- ggplot(plot_data, aes(x = tmp_celltype, y = !!sym(gene))) +
        geom_point(mapping = aes(color = perturbation, fill = perturbation), size = 0.35, stroke = 0.2, shape = 21, position = position_jitterdodge(dodge.width = 0.9)) +
        geom_violin(
            mapping = aes(color = dummy),
            position = position_dodge(width = 0.9),
            fill = NA,
            adjust = 1, trim = TRUE,
            scale = "width", show.legend = F,
            width = 0.72 # adjust width so the violine does not stick out too much
        ) +
        theme_Publication_side_legend() %+%
        theme(
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            legend.text = element_markdown(size = 15),
            legend.title = element_text(size = 16)
        ) +
        ylab("Expression\n(logcounts)") +
        guides(
            color = guide_legend(
                title = "Condition",
                override.aes = list(size = 2, alpha = 1, shape = NULL),
                hjust = 0
            ),
            fill = "none"
        ) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.08))) +
        stat_summary(fun = "mean", geom = "point", aes(group = perturbation), color = "black", size = 3, shape = 8, stroke = 0.95, position = position_dodge(width = 0.9)) +
        scale_color_manual(
            values = c("ctrl" = color_mapping[1], "NotchRNAi" = color_mapping[4], NotchCphRNAi = "#00ba38", A = "black", B = "black", C = "black"),
            labels = c(
                ctrl = "Control",
                NotchRNAi = "*Notch*<sup><i>RNAi</i></sup>",
                NotchCphRNAi = "*Cph*<sup><i>RNAi</i></sup>+*Notch*<sup><i>RNAi</i></sup>"
            )
        ) +
        scale_fill_manual(values = alpha(c(color_mapping[1], color_mapping[4], "#00ba38"), 0.4)) +
        ggtitle(paste(population, gene)) +
        scale_y_continuous(expand = expansion(mult = c(0, 0.15)))


    return(p)
}

significance_marker <- function(input) {
    if (input < 0.001) {
        return("***")
    } else if (input < 0.01) {
        return("**")
    } else if (input < 0.05) {
        return("*")
    } else {
        return("")
    }
}


for (g in overlap_diff_expr_nanodam) {
    print(g)
    population <- "progenitors"

    # Create an annotation dataframe
    max_ <- FetchData(progenitors, g)[, 1] %>% max()

    sig_notch_ctrl <- MAST_notch_ctrl %>%
        filter(gene == g) %>%
        pull(p_adj) %>%
        significance_marker(.)

    sig_notch_cphnotch <- MAST_notch_cphnotch %>%
        filter(gene == g) %>%
        pull(p_adj) %>%
        significance_marker(.)

    print(sig_notch_cphnotch)
    print(sig_notch_ctrl)
    annotation_df <- data.frame(
        celltype = c("a", "b"),
        y_position_ = c(max_ + 0.4, max_ + 0.8), # Adjust as needed
        xmin_ = c(0.7, 1),
        xmax_ = c(1, 1.3),
        label = c(sig_notch_ctrl, sig_notch_cphnotch)
    )

    p <- jitter_plot_fun(progenitors, g, population) +
        ggsignif::geom_signif(
            data = annotation_df,
            aes(xmin = xmin_, xmax = xmax_, annotations = label, y_position = y_position_, group = celltype),
            textsize = 6.5, manual = TRUE, vjust = 0,
            inherit.aes = F
        )

    ggsave(
        plot = p,
        filename = here("plots", "genes_under_cph_control", str_interp("${g}_jitter_plot_cph_neg_regulation.pdf")),
        width = 3.5,
        height = 1.8,
        scale = 2
    )
}
```




```{r}
sessionInfo()
```
